aten/src/ATen/native/cuda/TensorShape.cu

```diff
@@ -0,0 +1,399 @@
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
+#include <ATen/Dispatch.h>
+#include <ATen/core/Tensor.h>
+#include <ATen/cuda/CUDAContext.h>
+#include <ATen/ops/empty.h>
+#include <c10/cuda/CUDAGuard.h>
+#include <c10/util/Logging.h>
+
+namespace at::native {
+
+namespace detail {
+
+// NOTE [CUDA fast path for split_with_sizes_copy.out]
+// split_with_sizes_copy.out for contiguous operands has the following
+// properties:
+// - Each src split consists of multiple chunks that are separated by a fixed
+// stride. The number of chunks and the strides are the same across all src
+// splits.
+// - Each dst split is the concatenation of the chunks in its corresponding src
+// splits.
+// - The sizes chunks vary across splits.
+// - The alignment varies across chunks. A (src, dst) chunk pair is not
+// guaranteed to have the same alignment.
+//
+// The following strategies are employed to optimize for this workload:
+// - The entire workload is fused into a single kernel to maximize I/O
+// throughput and minimize wave quantization.
+// - To account for both small and large chunk sizes, a "jagged grid" is used.
+// Each chunk is processed by one or more blocks depending on its size.
+// - Within each chunk, the region in which writes can be vectorized is
+// identified. Within this region, writes are always vectorized and reads are
+// oppurtunistically vectorized.
+static constexpr int64_t BLOCK_SIZE = 128;
+static constexpr int64_t BYTES_PER_THREAD = 16;
+
+static __host__ __device__ inline int64_t div_up(int64_t a, int64_t b) {
+  return (a + b - 1) / b;
+}
+
+template <typename T>
+__device__ inline void stream_load128(uint4& val, const T* addr) {
+  uint64_t low, high;
+#if defined(USE_ROCM) || (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ < 800))
+  low = reinterpret_cast<const uint64_t*>(addr)[0];
+  high = reinterpret_cast<const uint64_t*>(addr)[1];
+#else
+  asm("ld.global.nc.v2.u64 {%0, %1}, [%2];"
```

Comment:
I was curious why we do not do `ld.global.cs.nc.v2.u64 ...` with the added `.cs` cache operation if this is a "stream" load128.